src/main/java/org/opensearch/neuralsearch/query/AgenticSearchQueryBuilder.java

```diff
@@ -0,0 +1,273 @@
+
+    public static void initialize(MLCommonsClientAccessor mlClient) {
+        AgenticSearchQueryBuilder.ML_CLIENT = mlClient;
+    }
+
+    // For testing purposes
+    static void resetMLClient() {
+        AgenticSearchQueryBuilder.ML_CLIENT = null;
+    }
+
+    public AgenticSearchQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        this.queryText = in.readString();
+        this.agentId = in.readString();
+        this.queryFields = in.readOptionalStringList();
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeString(this.queryText);
+        out.writeString(this.agentId);
+        out.writeOptionalStringCollection(this.queryFields);
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder xContentBuilder, Params params) throws IOException {
+        xContentBuilder.startObject(NAME);
+        if (Objects.nonNull(QUERY_TEXT_FIELD)) {
+            xContentBuilder.field(QUERY_TEXT_FIELD.getPreferredName(), queryText);
+        }
+        if (Objects.nonNull(AGENT_ID)) {
+            xContentBuilder.field(AGENT_ID.getPreferredName(), agentId);
+        }
+        if (Objects.nonNull(queryFields) && !queryFields.isEmpty()) {
+            xContentBuilder.field(QUERY_FIELDS.getPreferredName(), queryFields);
+        }
+        xContentBuilder.endObject();
+    }
+
+    /**
+     * Creates AgenticSearchQueryBuilder from XContent.
+     * The expected parsing form looks like:
+     * {
+     *  "agentic": {
+     *    "query_text": "string",
+     *    "agent_id": "string",
+     *    "query_fields": ["string", "string"..]
+     *    }
+     * }
+     *
+     */
+
+    public static AgenticSearchQueryBuilder fromXContent(XContentParser parser) throws IOException {
+        AgenticSearchQueryBuilder agenticSearchQueryBuilder = new AgenticSearchQueryBuilder();
+
+        String currentFieldName = null;
+        XContentParser.Token token;
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentFieldName = parser.currentName();
+            } else if (token.isValue()) {
+                if (QUERY_TEXT_FIELD.match(currentFieldName, parser.getDeprecationHandler())) {
+                    agenticSearchQueryBuilder.queryText = parser.text();
+                } else if (AGENT_ID.match(currentFieldName, parser.getDeprecationHandler())) {
+                    agenticSearchQueryBuilder.agentId = parser.text();
+                } else {
+                    throw new ParsingException(parser.getTokenLocation(), "Unknown field [" + currentFieldName + "]");
+                }
+            } else if (token == XContentParser.Token.START_ARRAY) {
+                if (QUERY_FIELDS.match(currentFieldName, parser.getDeprecationHandler())) {
+                    List<String> fieldsList = new ArrayList<>();
```

Comment:
we need to have a limit for number of fields we support, something big enough for potential use cases, maybe 25?

Reply:
Added the limit of 25 and a test